Implementation of ICub device

lrapetti opened the pull request:
Cleaned PR starting from https://github.com/robotology-playground/wearables/pull/16.

For the moment I am leaving the original commits, in case we may prefere to squash them.

yeshasvitirupachuri committed:
Icub device initial commit

yeshasvitirupachuri committed:
Implement sensor methods

yeshasvitirupachuri committed:
Implement getTimeStamp method

yeshasvitirupachuri committed:
Update CMakeLists.txt

yeshasvitirupachuri committed:
Update timestamp implementation

yeshasvitirupachuri committed:
Update CMakeLists and add config file

yeshasvitirupachuri committed:
Implement getLastInputStamp method

yeshasvitirupachuri committed:
Implement getStatus method

yeshasvitirupachuri committed:
Fix sensor type

yeshasvitirupachuri committed:
Update wrapper details in config

yeshasvitirupachuri committed:
Update status handling

yeshasvitirupachuri committed:
Implement bufferport callback

yeshasvitirupachuri committed:
Improve status update logic

yeshasvitirupachuri committed:
Remode debug code

yeshasvitirupachuri committed:
Update wrench handling

yeshasvitirupachuri committed:
Add mutex to wrench handling

yeshasvitirupachuri committed:
Add ICub config file

yeshasvitirupachuri committed:
Update PR changes

yeshasvitirupachuri committed:
Cleanup open configuration

yeshasvitirupachuri committed:
Update port connection

yeshasvitirupachuri committed:
Update separator usage

lrapetti committed:
remove Yarp Init check in ICub

yeshasvitirupachuri committed:
Update PR changes

yeshasvitirupachuri committed:
Beautify using wearables clang format file

yeshasvitirupachuri committed:
update config file names and add yarpmanager files

diegoferigo commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,345 @@
+    }
+
+    // Return a shared point to the required sensor
+    return static_cast<std::shared_ptr<sensor::IForceTorque6DSensor>>(pImpl->ftSensorsMap.at(name));
```
Since getting sensors is not done at runtime, probably is safer to use [`dynamic_pointer_cast`](http://www.cplusplus.com/reference/memory/dynamic_pointer_cast/). Furthermore, use the typedef of `SensorPtr<const wearable::sensor::IForceTorque6DSensor>` instead of `shared_ptr`. In this way, if for any reason we want to change what `SensorPtr` is, we just have to change the typedef.

Yeshasvitvs replied:
Done

diegoferigo replied:
Thinking more about this, since the map always contains an object which inherits from `IForceTorque6DSensor`, probably the cast is not strictly required. Can you try to remove it and if it complains about const correctness use [`const_pointer_cast`](http://www.cplusplus.com/reference/memory/const_pointer_cast/)?

lrapetti replied:
this was done in https://github.com/robotology/wearables/commit/c0b229ef04c9a2a09b0e073af326cb381dae60d3

traversaro commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,345 @@
+    // Configure clock
+    if (!yarp::os::Time::isSystemClock()) {
+        yarp::os::Time::useSystemClock();
+    }
```
Yes, It seems fishy also to me. Why is that used?

Yeshasvitvs replied:
I agree on the usage of network clock in case of using icub in gazebo! Ideally, at this stage, this device should connect to the wholebodydynamics ports at the hands end-effectors and pass the FT values as a wearable device. So, we can remove the explicit configuration of this device to the system clock.

diegoferigo replied:
@Yeshasvitvs Has this been fixed?

lrapetti replied:
done in https://github.com/robotology/wearables/pull/25/commits/42eba1b291ebfda8a27629c4a999cf1225bdca30

lrapetti replied:
done in https://github.com/robotology/wearables/pull/25/commits/42eba1b291ebfda8a27629c4a999cf1225bdca30

Yeshasvitvs commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,345 @@
+    std::vector<double> wrenchValues;
+    bool firstRun = true;
+    mutable std::mutex mtx;
+
```
Using the Network member was throwing compile-time error on Mac OS. So, it was removed in https://github.com/robotology-playground/wearables/commit/c0669c49ab749f81f81921fcf49ec322675dfd49. @lrapetti please add the details if you have any error messages.

lrapetti replied:
I removed that part in c0669c4 because otherwise in `MacOS` we were getting the following error:
```
[ 86%] Linking CXX shared module ../../lib/yarp/ICub.so
Undefined symbols for architecture x86_64:
  "yarp::os::Network::init()", referenced from:
      wearable::devices::ICub::open(yarp::os::Searchable&) in ICub.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/yarp/ICub.so] Error 1
make[1]: *** [devices/ICub/CMakeFiles/ICub.dir/all] Error 2
make: *** [all] Error 2
```

If I try to add a `yarp::os::Network` member, again I fall in similar error:
```
[ 86%] Linking CXX shared module ../../lib/yarp/ICub.so
Undefined symbols for architecture x86_64:
  "yarp::os::Network::init()", referenced from:
      wearable::devices::ICub::open(yarp::os::Searchable&) in ICub.cpp.o
  "yarp::os::Network::Network()", referenced from:
      WrenchPort::WrenchPort() in ICub.cpp.o
  "yarp::os::Network::~Network()", referenced from:
      WrenchPort::~WrenchPort() in ICub.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/yarp/ICub.so] Error 1
make[1]: *** [devices/ICub/CMakeFiles/ICub.dir/all] Error 2
make: *** [all] Error 2
```

diegoferigo replied:
I suspect that something was not linked correctly, can you try to link  `YARP::YARP_init` target?

lrapetti replied:
> I suspect that something was not linked correctly, can you try to link `YARP::YARP_init` target?

linking `YARP_YARP_init` it compiles correctly when adding `yarp::os::Network yarpNetwork;`

diegoferigo replied:
Great!

lrapetti replied:
@Yeshasvitvs if you want I can push this change

lrapetti replied:
Done with https://github.com/robotology-playground/wearables/pull/25/commits/ee3990a7328856798b72dbcc790f4eee7e4255b0.

Yeshasvitvs commented on devices/ICub/include/ICub.h:
```diff
@@ -0,0 +1,208 @@
+ * GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#ifndef ICUB_H
```
Done

Yeshasvitvs commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,345 @@
+        }
+        default: {
+            // yWarning() << LogPrefix << "Selected sensor type (" << static_cast<int>(type)
+            //           << ") is not supported by ICub";
```
I will uncomment this massage as it will give an user correct info related to the sensors user is trying to access through the configuration file.

Yeshasvitvs replied:
Removed this line as it is unnecessarily long warning message!

Yeshasvitvs replied:
Done

Yeshasvitvs commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,345 @@
+
+    while (pImpl->leftHandFTPort.firstRun) {
+        pImpl->leftHandFTPort.useCallback();
+    }
```
Done

yeshasvitirupachuri committed:
Update PR changes

lrapetti committed:
Add yarp network member to ICub

diegoferigo commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,343 @@
+    pImpl->rightHandFTPort.useCallback();
+
+    // Wait to receive first data
+    while (pImpl->leftHandFTPort.firstRun && pImpl->rightHandFTPort.firstRun) {
```
This should wait that both ports received the first data. With `&&` the flow continues when just one did. Using `||` would solve it.

diegoferigo replied:
@Yeshasvitvs Still TODO

lrapetti replied:
This comment seems to be outdated.

diegoferigo replied:
If the code changed, we can resolve this conversation

diegoferigo reviewed (changes requested):
Almost done in my opinion beyond previous unresolved comments. Remember to pass this file through clang-format before merging.

yeshasvitirupachuri committed:
Add and Implement VirtualJointKinSensor

yeshasvitirupachuri committed:
Update devices with VirtualJointKinSensor

Yeshasvitvs commented:
> Why not using the `RemoteControlBoardRemapper`? This would mean that here a controlled joint list should appear, and 1) in this way you avoid handling individual control boards and 2) it would allow to stream only a subset of joint measurements instead of all joints.

If I understood well, the `RemoteControlBoardRemapper` will identify all the control boards present without the need to explicitly state the control board names in the config file?

Yeshasvitvs commented:
> Since this can be used for generalized joints, I would call it `m_position`.

Done

Yeshasvitvs commented:
> Since this can be used for generalized joints, I would call it `m_position`.

Done

Yeshasvitvs commented:
> `SensorPtr`

Done

Yeshasvitvs commented:
> > Why not using the `RemoteControlBoardRemapper`? This would mean that here a controlled joint list should appear, and 1) in this way you avoid handling individual control boards and 2) it would allow to stream only a subset of joint measurements instead of all joints.
> 
> If I understood well, the `RemoteControlBoardRemapper` will identify all the control boards present without the need to explicitly state the control board names in the config file?

Following the discussion with @diegoferigo we can change to using `RemoteControlBoardRemapper` at a later stage and stick with `remote_controlboard` for the time being.

diegoferigo commented:
I did a mess with these comments, I forgot that this PR was open and instead of doing a review I  commented the file of the commit :sweat_smile: Anyway:

> If I understood well, the RemoteControlBoardRemapper will identify all the control boards present without the need to explicitly state the control board names in the config file?

You avoid using the pair control board index + joint index in the control board. You just use the joint name. Simple clear.

> Following the discussion with @diegoferigo we can change to using RemoteControlBoardRemapper at a later stage and stick with remote_controlboard for the time being.

Before the merge, possibly ;)

yeshasvitirupachuri committed:
Implement VirtualJointLinkSensor in ICub device

Update position variable name in VirtualJointKinSensor

Update interface pointer handling in joint sensor constructor

Update spherical joint variables name in XsensSuit device

lrapetti commented:
What is the status of this PR? I think it is going to get priority since the current master of [human-dynamics-estimation](https://github.com/robotology/human-dynamics-estimation) is based on this branch

diegoferigo commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,345 @@
+    bool firstRun = true;
+    mutable std::mutex mtx;
+
+    using yarp::os::BufferedPort<yarp::os::Bottle>::onRead;
```
cc @Yeshasvitvs @lrapetti

lrapetti replied:
if I try to switch here to `using yarp::os::BufferedPort<yarp::sig::Vector>::onRead;`, I am getting the following:
```
[ 84%] Building CXX object devices/ICub/CMakeFiles/ICub.dir/src/ICub.cpp.o
/Users/lrapetti/robotology-playground/wearables/devices/ICub/src/ICub.cpp:47:11: error: 
      using declaration refers into
      'yarp::os::BufferedPort<yarp::sig::Vector>::', which is not a base class
      of 'WrenchPort'
    using yarp::os::BufferedPort<yarp::sig::Vector>::onRead;
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [devices/ICub/CMakeFiles/ICub.dir/src/ICub.cpp.o] Error 1
make[1]: *** [devices/ICub/CMakeFiles/ICub.dir/all] Error 2
make: *** [all] Error 2

```

diegoferigo commented:
I pinged @lrapetti and @Yeshasvitvs  on few comments that haven't been addressed yet. If someone has the bandwidth, switching to the remote control board remapper would be amazin (I suspect that otherwise no one else will do it in the close future), but it is not a blocking feature.

yeshasvitirupachuri committed:
Update wrench handling through configuration file

yeshasvitirupachuri committed:
Implement virtual link sensors - link pose interface

yeshasvitirupachuri committed:
Add link velocity computation

diegoferigo commented on devices/CMakeLists.txt:
```diff
@@ -8,3 +8,8 @@ add_subdirectory(IAnalogSensorToIWear)
   add_subdirectory(XsensSuit)
 endif()
+
+if(iDynTree_DIR)
```
What is this variable? Maybe `iDynTree_FOUND`?  @Yeshasvitvs @lrapetti

lrapetti replied:
done in https://github.com/robotology/wearables/pull/25/commits/32747cb9a99ee46d04b583a0dde3d39f36a7f40e

diegoferigo commented on devices/ICub/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+    YARP::YARP_dev
+    YARP::YARP_init
+    iDynTree::idyntree-model
+    iDynTree::idyntree-high-level)
```
I think that iDynTree libraries can become `PRIVATE`

lrapetti replied:
done in https://github.com/robotology/wearables/pull/25/commits/32747cb9a99ee46d04b583a0dde3d39f36a7f40e

diegoferigo commented on devices/ICub/src/ICub.cpp:
```diff
@@ -0,0 +1,1067 @@
+#include <iDynTree/ModelIO/ModelLoader.h>
+#include <iDynTree/KinDynComputations.h>
+
+#include <assert.h>
```
cassert

lrapetti replied:
don in https://github.com/robotology/wearables/pull/25/commits/42eba1b291ebfda8a27629c4a999cf1225bdca30

diegoferigo reviewed (changes requested):
Minor comments for now, I still have to go through one last time ICub.cpp

lrapetti committed:
remove output port for virtualJointKinSensors

lrapetti committed:
fix ICub device CMake

lrapetti committed:
minor fixies on ICub device

lrapetti commented:
All the checks are passed.

> > Following the discussion with @diegoferigo we can change to using RemoteControlBoardRemapper at a later stage and stick with remote_controlboard for the time being.
> 
> Before the merge, possibly ;)

This is the only change request still pending.

@Yeshasvitvs @diegoferigo, if you agree I would proceed with merging so that we can unlock other issues.

cc @traversaro @kouroshD

traversaro commented:
I think @diegoferigo is not responsive  on GitHub and he already expressed in person his ok with merging the PR, so if you need a quick feedback you may need to contact him via chat.

diegoferigo commented:
I'm off this week and I doubt I can do a last pass on this PR before my return. As said last time, if the remote control board remapper is postponed, this PR is good to go. Please open an issue to keep track of this task.

If there are any questions please be specific and point me to the code so thatI can have a look with the phone.

Yeshasvitvs commented:
@lrapetti ok for me

lrapetti commented:
> I'm off this week and I doubt I can do a last pass on this PR before my return. As said last time, if the remote control board remapper is postponed, this PR is good to go. Please open an issue to keep track of this task.
> 

issue opened at https://github.com/robotology/wearables/issues/49
